File: typescript-workshop/part-4-behavior/tests/helper/InMemoryDiscountCodeRepository.ts
import { DiscountCode } from '../../src/domain/DiscountCode';
import { DiscountCodeRepository } from '../../src/repository/DiscountCodeRepository';

export class InMemoryDiscountCodeRepository implements DiscountCodeRepository {
  private readonly usedCodes = new Set<string>();

  private key(customerId: string, discountCode: DiscountCode): string {
    return customerId + ':' + discountCode;
  }

  checkDiscountCode(customerId: string, discountCode: DiscountCode): boolean {
    return this.usedCodes.has(this.key(customerId, discountCode));
  }

  markAsUsed(customerId: string, discountCode: DiscountCode): void {
    this.usedCodes.add(this.key(customerId, discountCode));
  }

  withUsedCode(customerId: string, discountCode: DiscountCode): this {
    this.usedCodes.add(this.key(customerId, discountCode));
    return this;
  }

  isMarkedAsUsed(customerId: string, discountCode: DiscountCode): boolean {
    return this.usedCodes.has(this.key(customerId, discountCode));
  }
}

File: csharp-workshop/Part4.Behavior.Tests/Helper/InMemoryDiscountCodeRepository.cs
using Part4.Behavior.Domain;
using Part4.Behavior.Repository;

namespace Part4.Behavior.Tests.Helper;

public class InMemoryDiscountCodeRepository : IDiscountCodeRepository
{
    private readonly HashSet<string> _usedCodes = new();

    public InMemoryDiscountCodeRepository WithUsedCode(string customerId, DiscountCode discountCode)
    {
        _usedCodes.Add(Key(customerId, discountCode));
        return this;
    }

    public bool IsMarkedAsUsed(string customerId, DiscountCode discountCode)
    {
        return _usedCodes.Contains(Key(customerId, discountCode));
    }

    public bool CheckDiscountCode(string customerId, DiscountCode discountCode)
    {
        return _usedCodes.Contains(Key(customerId, discountCode));
    }

    public void MarkAsUsed(string customerId, DiscountCode discountCode)
    {
        _usedCodes.Add(Key(customerId, discountCode));
    }

    private static string Key(string customerId, DiscountCode discountCode) =>
        $"{customerId}:{discountCode}";
}
